rust/kernel/platform_driver.rs

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::{
+    bindings, c_types,
+    error::{Error, Result},
+    types::PointerWrapper,
+    CStr,
+};
+use alloc::boxed::Box;
+use core::{marker::PhantomPinned, mem::transmute, pin::Pin};
+
+extern "C" {
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_get_drvdata(
+        pdev: *const bindings::platform_device,
+    ) -> *mut c_types::c_void;
+
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_set_drvdata(
+        pdev: *mut bindings::platform_device,
+        data: *mut c_types::c_void,
+    );
+}
+
+unsafe extern "C" fn probe_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let f = || {
+        let drv_data = T::probe(&mut PlatformDevice::new(pdev))?;
+        let drv_data = drv_data.into_pointer() as *mut c_types::c_void;
+        Ok(drv_data) as Result<_>
+    };
+    let ptr = match f() {
```

Comment:
It seems a bit odd stylistically to use a function for this, I think there's a macro in `file_operations` that might be more what you want here?